flocker/acceptance/endtoend/test_dataset.py

```diff
@@ -35,6 +40,112 @@ def test_dataset_creation(self, cluster):
         """
         return create_dataset(self, cluster)
 
+    def _get_package_source(self, default_version=None):
+        """
+        Get the package source for the flocker version under test from
+        environment variables.
+
+        :param unicode default_version: The version of flocker to use
+            if not specified in environment variables.
+
+        :return: A ``PackageSource`` that can be used to install the version of
+            flocker under test.
+        """
+        env_vars = ['FLOCKER_ACCEPTANCE_PACKAGE_BRANCH',
+                    'FLOCKER_ACCEPTANCE_PACKAGE_VERSION',
+                    'FLOCKER_ACCEPTANCE_PACKAGE_BUILD_SERVER']
+        missing_vars = list(x for x in env_vars if x not in os.environ)
+        if missing_vars:
+            raise SkipTest(
+                'Missing environment variables for upgrade test: %s.' %
+                ', '.join(missing_vars))
+        version = (os.environ['FLOCKER_ACCEPTANCE_PACKAGE_VERSION'] or
+                   default_version)
+        return PackageSource(
+            version=version,
+            branch=os.environ['FLOCKER_ACCEPTANCE_PACKAGE_BRANCH'],
+            build_server=os.environ['FLOCKER_ACCEPTANCE_PACKAGE_BUILD_SERVER'])
+
+    @skip_backend(
+        unsupported={DatasetBackend.loopback},
+        reason="Does not maintain compute_instance_id across restarting "
+               "flocker (and didn't as of most recent release).")
+    @run_test_with(async_runner(timeout=timedelta(minutes=6)))
+    @require_cluster(1)
+    def test_upgrade(self, cluster):
+        """
+        Given a dataset created and used with the previously installable
+        version of flocker, uninstalling the previous version of flocker and
+        installing HEAD does not destroy the data on the dataset.
+        """
+        node = cluster.nodes[0]
+        SAMPLE_STR = '123456' * 100
+
+        upgrade_from_version = get_installable_version(version)
+
+        # Get the initial flocker version and setup a cleanup call to restore
+        # flocker to that version when the test is done.
+        d = cluster.client.version()
+        original_package_source = [None]
+
+        def setup_restore_original_flocker(version):
+            original_package_source[0] = (
+                self._get_package_source(
+                    default_version=str(version.get('flocker')) or None)
```

Comment:
Unfortunately `version` is parsed JSON, so it's more like `dict(flocker=u"1.9.0.dev.asdsadsa")`

And PackageSource requires version to be of type bytes. Switched to explicitly 'encoding' the unicode, and corrected a bug here.